src/serialization.ts

```diff
@@ -0,0 +1,212 @@
+        throw new Error('No signatures were provided');
+    }
+
+    const putSignature = (signature: string) => {
+        const signatureBytes = Buffer.from(signature, 'hex');
+        const length = Buffer.alloc(2);
+        length.writeUInt16BE(signatureBytes.length, 0);
+        return Buffer.concat([length, signatureBytes]);
+    };
+    const putCredentialSignatures = (credSig: CredentialSignature) =>
+        serializeMap(credSig, encodeUint8, encodeUint8, putSignature);
+    return serializeMap(
+        signatures,
+        encodeUint8,
+        encodeUint8,
+        putCredentialSignatures
+    );
+}
+
+/**
+ * Serializes a transaction and its signatures. This serialization when sha256 hashed
+ * is considered as the transaction hash, and is used to look up the status of a
+ * submitted transaction.
+ * @param accountTransaction the transaction to serialize
+ * @param signatures signatures on the signed digest of the transaction
+ * @returns the serialization of the account transaction, which is used to calculate the transaction hash
+ */
+export function serializeAccountTransaction(
+    accountTransaction: AccountTransaction,
+    signatures: AccountTransactionSignature
+): Buffer {
+    const serializedBlockItemKind = encodeUint8(
+        BlockItemKind.AccountTransactionKind
+    );
+    const serializedAccountTransactionSignatures =
+        serializeAccountTransactionSignature(signatures);
+
+    const serializedType = serializeAccountTransactionType(
+        accountTransaction.type
+    );
+
+    const accountTransactionHandler = getAccountTransactionHandlerMap().get(
+        accountTransaction.type
+    );
+    const serializedPayload = accountTransactionHandler.serialize(
+        accountTransaction.payload
+    );
+
+    const baseEnergyCost = accountTransactionHandler.getBaseEnergyCost(
+        accountTransaction.payload
+    );
+    const energyCost = calculateEnergyCost(
+        countSignatures(signatures),
+        BigInt(serializedPayload.length + 1),
+        baseEnergyCost
+    );
+    const serializedHeader = serializeAccountTransactionHeader(
+        accountTransaction.header,
+        serializedPayload.length + 1,
+        energyCost
+    );
+
+    return Buffer.concat([
+        serializedBlockItemKind,
+        serializedAccountTransactionSignatures,
+        serializedHeader,
+        serializedType,
+        serializedPayload,
+    ]);
+}
+
+/**
+ * Gets the transaction hash that is used to look up the status of a transaction.
+ * @param accountTransaction the transaction to hash
+ * @param signatures the signatures that will also be part of the hash
+ * @param sha256 the sha256 hashing function
+ * @returns the sha256 hash of the serialized block item kind, signatures, header, type and payload
+ */
+export function getAccountTransactionHash(
+    accountTransaction: AccountTransaction,
+    signatures: AccountTransactionSignature,
+    sha256: (serialization: Buffer[]) => Buffer
```

Comment:
The reason is also quite convoluted. In the desktop wallet we like to do serialization on the renderer threads, as they are used quite a bit in the UI. In a renderer thread node modules are not directly available, and the crypto library is one of these. Providing the function as a parameter means that we can avoid a lot of IPC Electron async magic when we start using this in the desktop wallet. An alternative would be to find a non-node sha256 implementation.

Reply:
Ahh that make sense. Unless we case about internet explorer, every browser should have sha builtin [SubtleCrypto](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest)

Reply:
I have refactored it so that we provide the sha256 function to make the interface cleaner. We can always revisit if the used package for sha256 shows to be incompatible for running on the renderer threads.